### DISCUSSION TITLE
System.Linq.Enumerable: Add `IsNullOrEmpty()`

### DISCUSSION THREAD

**Discussion by tvardero** [dotnet/runtime#77227] [Ideas]
_Originally posted here: https://github.com/dotnet/csharplang/discussions/6572_

Proposing to add this to System.Linq.Enumerable class:
```c#
public bool IsNullOrEmpty([NotNullWhen(false)] IEnumerable? value);
```

Idea behind: having `string.IsNullOrEmpty()` for everything enumerable.

Could come in handy:
```c#
IEnumerable<object>? nullCollection = (List<object>)null;
IEnumerable<object>? emptyCollection = Array.Empty<object>();
IEnumerable<object>? notEmptyCollection = (IEnumerable<object>)(IEnumerable<char>)"something";

// Existing solution:
bool a1 = nullCollection == null || !nullCollection.Any(); // true
bool b1 = emptyCollection == null || !emptyCollection.Any(); // true
bool c1 = notEmptyCollection = null || !notEmptyCollection.Any(); // false

// Proposed:
bool a2 = Enumerable.IsNullOrEmpty(nullCollection); // true
bool b2 = Enumerable.IsNullOrEmpty(emptyCollection); // true
bool c2 = Enumerable.IsNullOrEmpty(notEmptyCollection); // false

// Also speaking of System.String class:
_ = string.IsNullOrEmpty("something") == Enumerable.IsNullOrEmpty("something");
_ = string.IsNullOrWhiteSpace("something") == Enumerable.IsNullOrEmpty("something".Trim());
```

**Comment by EgorBo** [comment]
Related: https://github.com/dotnet/runtime/issues/32642

**Reply by tvardero** [reply]
Thank you. As we know, Array also is IEnumerable, so my proposal also covers those. As well as String, as latter is just IEnumerable of chars with some salt and pepper.

I don't propose, however, to change all existing .Net codebase, because my solution does not actually change anything, except for adding sweet flavor to the code 😅.  
Just could be handy for consumers who write 100+ statements inside one `if` clause, I know few of those myself 😁.

**Comment by FaustVX** [comment]
This should be
```C#
public static bool IsNullOrEmpty([NotNullWhen(false)] this IEnumerable? value);
```
With the `static` and `this`

**Reply by tvardero** [reply]
You will get NullReferenceException on collections that are `null`.

**Reply by FaustVX** [reply]
No, that's not how extension methods works
[sharplab.io](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYAMBYAUAYiggBtiBDAI2IFMACaqSmvDARhc1tVYDoAZAJZQAjgG4OXXgGEA9qWoBjYAJlQAzuNypO3HgBEBZAOZQZa5QrU9ZcagEFGxAJ5qBGidwAsmvIPMAeGQoAK0VgAD4AfloAN1oAXloiUh8tVgBOAAoYngBJNQA5EmIAeTAAUQBbAAdgJ0yASgbU7gA2LnRacrwAbzxaAa4AZkl2ihk5Wnyi0jKq2vqAbQKZYBniAHUACwZMgDMyYjVqBoBdWmAtt0kh6JjDiBP+wZf48NiHumvk4k0AXyAA=)

**Reply by tvardero** [reply]
Wow, more I live, more I learn. Never thought that I could use extensions method invocations on something that is currently `null` 😅
_(Yep, because extension method do not belong to the object they are being invoked on. They are static methods of some other class.)_

However the reason I did so, is because I want it be similar to already existing `string.IsNullOrEmpty()`. It would be much more expected / intuitive to developers because they saw something similar already.

**Reply by FaustVX** [reply]
Yes, but all `IEnumerable` methods are `static`

**Reply by bartonjs** [reply]
There's a guideline that says not to allow `null` to work in (public) extension methods, even though the language/runtime allows it; because it definitely confuses people when `thingIKnowIsNull.Something()` doesn't throw.  So an Enumerable.IsNullOrEmpty, if added, would be a normal/non-extension method.

**Reply by FaustVX** [reply]
Ok, it makes sense
